# ACM + Lyft HackNight - Python Starter Notebook

<strong>Prompt:</strong> In a city of only taxis from the Bauer Taxi Service, your team is put in charge to earn Lyft more business and win riders over. The ACM Executive team has provided you with this starter notebook, which showcases how you can utilize the RESTful API that has 5 public endpoints that you can make calls to. These endpoints are listed below:

<ul>
    <li><strong>/time:</strong> GET</li>
        <ul>
            <li>Returns the simulation period you are developing your pricing model for.</li>
        </ul>
    <li><strong>/trips:</strong> GET</li>
        <ul>
            <li>Returns the taxi trips that occured with the criterion specified.</li>
        </ul>
    <li><strong>/count:</strong> GET</li>
        <ul>
            <li>Returns the total amount of trips that occured during a certain time period.</li>
        </ul>
    <li><strong>/pricing:</strong> POST</li>
        <ul>
            <li>Submits your team's general pricing strategy.</li>
        </ul>
    <li><strong>/zones:</strong> POST</li>
        <ul>
            <li>Submits the second component of your pricing strategy: the areas you believe are power zones.</li>
        </ul>
</ul>

<strong>Game Overview:</strong> There are 5 simple steps you need to follow to succeed this HackNight!
<ol>
    <li>Wait for a "simulation" (access to taxi data of a specific week and prior)</li>
    <li>Use the endpoints with associated GET requests to do data analysis to develop your pricing model.</li>
    <li>Submit your pricing model</li>
    <li>View your revenue generated (the results) on the diagram shown at the end of each simulation.</li>
    <li>Repeat this process for every simulation</li>
</ol>

## Learning Objective: Using a RESTful API

An API (Application Programming Interface) is a set of methods that controls the access a programmer has to what they want to reach (in this case, the taxi dataset). Here, the Lyft HackNight API limits access to taxi data by bounding it to the simulation. What does this mean? It means that you can only access taxi data of a certain date or date range when it is within the simulation date range or prior.

Now, let's get started with set up process.

### Initial Setup and Imports

In [4]:
# Required imports
import requests # Python's go-to HTTP request library
from datetime import datetime

In [5]:
# Environment and other variables for the API to know who you are
URL = "http://localhost:8080"
TEAM = "team4"

### Our first GET request: /time

Format of a GET request with Python's request library:

requests.get(SERVER_URL + "/ENDPOINT_NAME/", params="{QUERY HERE}).json()

In [13]:
requests.get(URL + "/time/").json()

{'message': 'The simulation is not over, the time is 10/1/2017.',
 'sucess': True,
 'time': 1506834000000}

Note that the /time endpoint doesn't take any parameters, so there's no need to specify "params" in the method call. However, in the /trips endpoint below, we'll do just that.

### /trips endpoint

The /trips endpoint allows us to find all trips that have occured in a specific time range. Note, however, that you can only search for trips that happen in the current simulation. We don't want you to predict the future!

<strong>Endpoint: /trips</strong>
<ul>
    <li>team (str): the string name of your team as indicated in the "TEAM" variable above</li>
    <li>start (str): the date and time you want to start searching from, in format %m/%d/%Y %H:%M A/PM</li>
        <ul>
            <li>Example: "10/07/2017 5:00PM"</li>
        </ul>
    <li>end (str): the date and time you want to end the search on, in format %m/%d/%Y %H:%M A/PM</li>
    <li>limit (int): maximum number of results to return</li>
    <li>offset (int): indicates you only want to receive data points past this index</li>
</ul>

In [32]:
# Build the query
trips_query = {
    "team": TEAM,
    "start": "9/10/2017 2:00 PM",
    "end": "9/10/2017 3:00 PM",
    "limit": 1
}

# Send the request
trips_response = requests.get(URL + "/trips/", params=trips_query).json()
trips_response

{'length': 1,
 'response': [{'company': 'Bauer Taxi Service',
   'dropoff_centroid_latitude': '41.877406123',
   'dropoff_centroid_longitude': '-87.621971652',
   'dropoff_community_area': '32',
   'entry_idx': 0,
   'extras': '0',
   'fare': '4.75',
   'pickup_centroid_latitude': '41.877406123',
   'pickup_centroid_longitude': '-87.621971652',
   'pickup_community_area': '32',
   'taxi_id': '52789c130fe640cb73a4081eca9872f696ab63daa15effcd8b74739e40609b524f1dad55ea7b820852285b3ebffe00b626a93fae3751e9086f6e1c74ccc76e91',
   'tips': '0',
   'tolls': '0',
   'trip_end_timestamp': '2017-09-10T09:15:00',
   'trip_id': '0007c73adceba3169364c89a8ad2b2073530acf3',
   'trip_miles': '0.5',
   'trip_seconds': '240',
   'trip_start_timestamp': '2017-09-10T09:15:00',
   'trip_total': '4.75'}],
 'success': True}

The JSON (Javascript Object Notation) object we got back from our call to /trips is 1 ride that occured from 2:00 - 3:00PM. In Python, this JSON object is interpreted as a dictionary. <strong>We care about the "response" key.</strong> To retrieve the response, we simply use Python's dictionary bracket notation, which will give us the datapoint we want.

In [27]:
# Gives us a list of the taxi rides that fits the criterion we wanted. Only one in here since we set limit = 1.
trips_response["response"]

[{'company': 'Bauer Taxi Service',
  'dropoff_centroid_latitude': '41.877406123',
  'dropoff_centroid_longitude': '-87.621971652',
  'dropoff_community_area': '32',
  'entry_idx': 0,
  'extras': '0',
  'fare': '4.75',
  'pickup_centroid_latitude': '41.877406123',
  'pickup_centroid_longitude': '-87.621971652',
  'pickup_community_area': '32',
  'taxi_id': '52789c130fe640cb73a4081eca9872f696ab63daa15effcd8b74739e40609b524f1dad55ea7b820852285b3ebffe00b626a93fae3751e9086f6e1c74ccc76e91',
  'tips': '0',
  'tolls': '0',
  'trip_end_timestamp': '2017-09-10T09:15:00',
  'trip_id': '0007c73adceba3169364c89a8ad2b2073530acf3',
  'trip_miles': '0.5',
  'trip_seconds': '240',
  'trip_start_timestamp': '2017-09-10T09:15:00',
  'trip_total': '4.75'}]

#### get_trips(query) method

Since your goal is to look at many different taxi rides, we want to simplify things by creating a method that just takes in the query we want to process. To do this, we simply put the repeated information inside of the method that *wraps* the the actual GET request.

In [28]:
def get_trips(query):
    query["team"] = TEAM
    response = requests.get(URL + "/trips/", params=query)
    return response.json()

In [57]:
# Example usage of get_trips - no need to specify the team or GET request anymore!
get_trips({
        "start": "9/10/2017 3:45 AM",
        "end": "9/10/2017 3:46 AM",
        "limit": 5
    })

{'length': 4,
 'response': [{'company': 'Bauer Taxi Service',
   'dropoff_centroid_latitude': '41.980264315',
   'dropoff_centroid_longitude': '-87.913624596',
   'dropoff_community_area': '76',
   'entry_idx': 0,
   'extras': '0',
   'fare': '43.5',
   'pickup_centroid_latitude': '41.980264315',
   'pickup_centroid_longitude': '-87.913624596',
   'pickup_community_area': '76',
   'taxi_id': 'a3bc50ed4e7b5de8e74d56919ae6e14b91e33c63c6c7d9632165c99dd1781507f170419466c56588531ed4788c84cabc34a62efd4834750e7ae2753c3667fc1d',
   'tips': '0',
   'tolls': '0',
   'trip_end_timestamp': '2017-09-09T23:00:00',
   'trip_id': '007bcf68e2b126700ca88588bc4531e86ddef63f',
   'trip_miles': '17.6',
   'trip_seconds': '1380',
   'trip_start_timestamp': '2017-09-09T22:45:00',
   'trip_total': '43.5'},
  {'company': 'Bauer Taxi Service',
   'dropoff_centroid_latitude': '41.892658108',
   'dropoff_centroid_longitude': '-87.652534484',
   'dropoff_community_area': '24',
   'entry_idx': 1,
   'extras': '1',


#### Sending SQL queries

The GET request for /trips can accept SQL queries when passes as strings.

<strong>Parameters:</strong>
<ul>
    <li>team</li>
    <li>where : Indicates the SQL WHERE clause which defines the query.</li>
    <li>limit</li>
</ul>

In [59]:
get_trips({
        "where": "(trip_start_timestamp BETWEEN {9/10/2017 2:00 PM} AND {9/10/2017 3:00 PM}) AND (trip_total BETWEEN 10 AND 20)",
        "limit": 1
})

{'length': 1,
 'response': [{'company': 'Bauer Taxi Service',
   'dropoff_centroid_latitude': '41.899602111',
   'dropoff_centroid_longitude': '-87.633308037',
   'dropoff_community_area': '8',
   'entry_idx': 0,
   'extras': '0',
   'fare': '13.25',
   'pickup_centroid_latitude': '41.878865584',
   'pickup_centroid_longitude': '-87.625192142',
   'pickup_community_area': '32',
   'taxi_id': '1095c5432cb9cfd88296fb166e38d560936d4690801dddfae2bbae88a2d295542548a58e8e2394de8c67c82610261281b4303ad7f0b85236f9e7c4f46badf414',
   'tips': '0',
   'tolls': '0',
   'trip_end_timestamp': '2017-09-10T09:15:00',
   'trip_id': '00197d0ae9d887315d40fb4a5a7405a062137c7d',
   'trip_miles': '3.9',
   'trip_seconds': '960',
   'trip_start_timestamp': '2017-09-10T09:00:00',
   'trip_total': '13.25'}],
 'success': True}

NOTE: Times are rounded off to every quarter hour. By specifying that you want the rides in between 3:00PM and 3:15PM, you're also getting back rides from 3:15PM to ~3:23PM. This is because the rides that happened at this time rounded down to 3:15PM.

You might be asking... well, how do we parse this data? We'll take a look at that in the data analysis section.

### /count endpoint

In [43]:
def get_count(query):
    query["team"] = TEAM
    response = requests.get(URL + "/count/", params=query)
    return response.json()

#### /count example : What if our simulation lands on New Year's?

In [55]:
# New Years
new_years_count = get_count({
        "start": "12/31/2016 12:00 AM",
        "end": "01/01/2017 05:00 AM"
    })
print("# of rides in 29-hour span of New Years Eve -> Day: {}".format(new_years_count["count"]))

# of rides in 29-hour span of New Years Eve -> Day: 74162


## Learning Objective: Data Preprocessing and Analysis

Before applying any data analysis to your data, it's always important to preprocess your data. This means looking for any missing values, identifying categorical variables, accounting for outlier (Y) or leverage (X) points. We'll go through the process of getting the data for a specific simulation, processing it with the stack we're familiar with (scipy), and begin some explatory data analysis on it.

In [62]:
# What data can we currently analyze?
requests.get(URL + "/time/").json()

{'message': 'The simulation is not over, the time is 10/1/2017.',
 'sucess': True,
 'time': 1506834000000}

Since the simulation time is 10/1/2017, we can look at data before 10/1/2017 12:00 AM. For this example analysis, we'll look at the week before the simulation time.

In [67]:
# How much data should we expect?
get_count({
        "start": "09/24/2017 12:00 AM",
        "end": "10/1/2017 12:00 AM"
    })

{'count': '342326', 'success': True}

In [72]:
# 1 week before the simulation time ends
get_trips({
        "start": "09/24/2017 12:00 AM",
        "end": "10/1/2017 12:00 AM"
    })["length"]

717

<strong>Well... that's odd.</strong> The call to /count is telling us that we should expect 342,326 taxi rides in the week of 09/24 - 10/01. However, we're only getting results back 717 trips back for the same date range call to /trips. This is because, on the backend, the Lyft RESTful API sets an internal <strong>limit</strong> of 1000 taxi rides.

This presents an interesting problem: how do we get back all of the taxi rides of a specific time range if we're bound to just 1000? Well, we have to use the offset paramater and set it to the length 1000 to get the next set of taxi rides. To get all 342,326 rides, we would have to do this process 342 times. This is simply unreasonable, so instead we need a good <strong>sampling strategy</strong>.

### Converting JSON to Pandas DataFrame

In [92]:
week_before = get_trips({
        "start": "09/24/2017 12:00 AM",
        "end": "10/1/2017 12:00 AM",
        "limit": 100
    })

In [93]:
week_before = week_before["response"]

In [96]:
import pandas as pd
import numpy as np

week_before_dataset = pd.DataFrame(week_before)
week_before_dataset

company dropoff_centroid_latitude dropoff_centroid_longitude  \
0   Bauer Taxi Service              41.906025969              -87.675311622   
1   Bauer Taxi Service              41.890922026              -87.618868355   
2   Bauer Taxi Service              41.934659157              -87.646729729   
3   Bauer Taxi Service              41.891971508              -87.612945414   
4   Bauer Taxi Service              41.835117986              -87.618677767   
5   Bauer Taxi Service              41.944226601              -87.655998182   
6   Bauer Taxi Service              41.892507781              -87.626214906   
7   Bauer Taxi Service              41.944226601              -87.655998182   
8   Bauer Taxi Service              41.884987192              -87.620992913   
9   Bauer Taxi Service              41.904935302              -87.649907226   
10  Bauer Taxi Service               41.92276062              -87.699155343   
11  Bauer Taxi Service              41.859349715              -87.617358006   
12  Bauer Taxi Service              41.899602111              -87.633308037   
13  Bauer Taxi Service              41.899602111              -87.633308037   
14  Bauer Taxi Service              41.879255084              -87.642648998   
15  Bauer Taxi Service               41.97907082              -87.903039661   
16  Bauer Taxi Service              41.877406123              -87.621971652   
17  Bauer Taxi Service              41.892072635              -87.628874157   
18  Bauer Taxi Service                41.9867118              -87.663416405   
19  Bauer Taxi Service              41.880994471              -87.632746489   
20  Bauer Taxi Service              41.901206994              -87.676355989   
21  Bauer Taxi Service              41.890922026              -87.618868355   
22  Bauer Taxi Service              41.901206994              -87.676355989   
23  Bauer Taxi Service              41.877406123              -87.621971652   
24  Bauer Taxi Service              41.878865584              -87.625192142   
25  Bauer Taxi Service              41.880994471              -87.632746489   
26  Bauer Taxi Service              41.898331794              -87.620762865   
27  Bauer Taxi Service              41.880994471              -87.632746489   
28  Bauer Taxi Service              41.785998518              -87.750934289   
29  Bauer Taxi Service               41.97907082              -87.903039661   
..                 ...                       ...                        ...   
45  Bauer Taxi Service              41.849246754              -87.624135298   
46  Bauer Taxi Service              41.884987192              -87.620992913   
47  Bauer Taxi Service              41.880994471              -87.632746489   
48  Bauer Taxi Service               41.79259236              -87.769615453   
49  Bauer Taxi Service              41.892507781              -87.626214906   
50  Bauer Taxi Service              41.870607372              -87.622172937   
51  Bauer Taxi Service                41.9867118              -87.663416405   
52  Bauer Taxi Service              41.899155613              -87.626210532   
53  Bauer Taxi Service              41.785998518              -87.750934289   
54  Bauer Taxi Service              41.880994471              -87.632746489   
55  Bauer Taxi Service              41.880994471              -87.632746489   
56  Bauer Taxi Service              41.879255084              -87.642648998   
57  Bauer Taxi Service              41.892072635              -87.628874157   
58  Bauer Taxi Service              41.901206994              -87.676355989   
59  Bauer Taxi Service              41.877406123              -87.621971652   
60  Bauer Taxi Service              41.884987192              -87.620992913   
61  Bauer Taxi Service              41.938391258               -87.63857492   
62  Bauer Taxi Service              41.898331794              -87.620762865   
63  Bauer Taxi Service              41.899602111              -87

### Finding the Average "fare"

Let's try to isolate the fare column so we can get the average. It's on index 6, and we can use the <strong>iloc</strong> method to capture it. The <strong>describe()</strong> function will give us summary statistics of the data.

In [106]:
# Let's isolate the "fare" col
week_before_dataset.iloc[:, 6].describe()

count       75
unique      51
top       7.75
freq         4
Name: fare, dtype: object

Well... that's odd. Where's the mean or median output? It turns out that the data returned to us in "fare" is in string format, so we need to convert the column to be a float.

In [131]:
fare_feature = week_before_dataset.iloc[:, [6]].copy()
fare_feature = fare_feature.astype("float64")
fare_feature.head()

fare
0  40.75
1   7.00
2   8.00
3  45.00
4  20.25

In [133]:
# Confirming that we've converted to a float
fare_feature.dtypes

fare    float64
dtype: object

#### Summary Statistics

In [134]:
fare_feature.describe()

fare
count  75.00000
mean   17.24000
std    14.89022
min     4.50000
25%     7.12500
50%     9.25000
75%    22.00000
max    66.50000

### Finding the Most Common Zone

Power zones are areas you can configure that require riders to pay an extra cost to take a Lyft, but with the benefit of it coming faster than, say, a taxi. The zone mappings can be seen in the <strong>"dropoff_community_area"</strong> feature. Let's try to find the most common zone, and then simply use that as the only power zone we want to apply for this simulation's pricing model.

In [136]:
week_before_dataset.iloc[:, 3].describe()

count     75
unique    15
top        8
freq      22
Name: dropoff_community_area, dtype: object

Let's take the average fare divided by 2 we found in the last section, coupled with the most frequent zone, and set those as the fares for the Lyft and power zones list, respectively!

In [138]:
def set_pricing(pricing):
    query = pricing;
    query["team"] = TEAM;
    response = requests.post(URL + "/pricing/", params=query)
    return response.json()

In [139]:
def set_zones(zones):
    zone_list = (",").join(str(z) for z in zones)
    query = {
        "team": TEAM,
        "zones": zone_list
    }
    response = requests.post(URL + "/zones/", params=query)
    return response.json()

In [140]:
my_price_model = set_pricing({
    "base": 17.24 / 2,
    "pickup": 0.00,
    "per_mile": 0.00,
    "per_minute": 0.00
})

In [141]:
my_power_zones = set_zones([8])